system/Database/Database.php

```diff
@@ -135,4 +141,50 @@ public function loadUtils(ConnectionInterface $db)
 	}
 
 	//--------------------------------------------------------------------
+
+	/**
+	 * Parse universal DSN string
+	 *
+	 * @param array $params
+	 *
+	 * @return array
+	 * @throws InvalidArgumentException
+	 */
+	protected function parseDSN($params): array
+	{
+		if (($dsn = parse_url($params['DSN'])) === false)
+		{
+			throw new \InvalidArgumentException('Your DSN connection string is invalid.');
+		}
+
+		$dsnParams = [
+			'DSN'      => '',
+			'DBDriver' => $dsn['scheme'],
+			'hostname' => isset($dsn['host']) ? rawurldecode($dsn['host']) : '',
+			'port'     => isset($dsn['port']) ? rawurldecode($dsn['port']) : '',
+			'username' => isset($dsn['user']) ? rawurldecode($dsn['user']) : '',
+			'password' => isset($dsn['pass']) ? rawurldecode($dsn['pass']) : '',
+			'database' => isset($dsn['path']) ? rawurldecode(substr($dsn['path'], 1)) : '',
+		];
+
+		// Do we have additional config items set?
+		if (! empty($dsn['query']))
+		{
+			parse_str($dsn['query'], $extra);
+
+			foreach ($extra as $key => $val)
+			{
+				if (is_string($val) && in_array(strtolower($val), ['true', 'false', 'null']))
```

Comment:
can use strict `in_array`:

```suggestion
                              if (is_string($val) && in_array(strtolower($val), ['true', 'false', 'null'], true))
```